File: delete_n_earn.py
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : A bit in python 


# // Your code here along with comments explaining your approach


class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        if len(nums)==0 or nums==None:
            return 0
        
        if len(nums)==1:
            return nums[0]
        

        # making an arry of max length of the nums array
        maximum=nums[0]
        for i in range(0,len(nums)):
            maximum=max(maximum,nums[i])
            
        
        #making the sum of each elemnt into their respective index
        dp=[0]*(maximum+1)
        for i in range(len(nums)):
                dp[nums[i]]+=nums[i]
        

        #same robber trick to find the max using 2 pointers
        prev=0
        curr=max(prev,dp[1])
        for i in range(2,len(dp)):
            temp=curr
            curr=max(curr,prev+dp[i])
            prev=temp
        
        return curr

File: DeleteNEarn.cpp
// Time Complexity - O(n)
// Space Complexity - O(n)
// Problems Faced - No!
// It runs on leetcode.

class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int m = 0;
        for(int i = 0; i < nums.size(); i++){
            m = max(m, nums[i]);
        }
        vector<int> nums2(m+1);
        
        for(int n : nums){
            nums2[n] += n;
        }
        
        vector<int> dp(m+1);
        dp[0] = nums2[0];
        dp[1] = max(nums2[0], nums2[1]);
        for(int i = 2; i < dp.size(); i++){
            dp[i] = max(dp[i-1], nums2[i] + dp[i-2]);
        }
        
        return dp[m];
        
    }
};
